src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp

```diff
@@ -453,6 +479,41 @@ REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalRegisterHijackCallback(_In_ PalH
 REDHAWK_PALEXPORT void REDHAWK_PALAPI PalHijack(HANDLE hThread, _In_opt_ void* pThreadToHijack)
 {
     _ASSERTE(hThread != INVALID_HANDLE_VALUE);
+
+#ifdef FEATURE_SPECIAL_USER_MODE_APC
+    if (g_pfnQueueUserAPC2Proc == QUEUE_USER_APC2_UNINITIALIZED)
+    {
+        g_pfnQueueUserAPC2Proc = (QueueUserAPC2Proc)GetProcAddress(LoadKernel32dll(), "QueueUserAPC2");
+    }
+
```

Comment:
This is not thread safe in C/C++.

It is valid for C/C++ compiler to rewrite this as:

```
QueueUserAPC2Proc pfn = g_pfnQueueUserAPC2Proc;
if (g_pfnQueueUserAPC2Proc == QUEUE_USER_APC2_UNINITIALIZED)
{
   g_pfnQueueUserAPC2Proc = pfn = (QueueUserAPC2Proc)GetProcAddress(LoadKernel32dll(), "QueueUserAPC2");
}

if (pfn)
{
....
}
```

It needs to use volatile or something like that.

Reply:
Only one thread may perform suspension/hijacking at a time. This code is single threaded, protected by the thread store lock.

Reply:
It needs comment at least to make this assumption explicit.

Reply:
I’ve made the variable ‘volatile’ initially n this change, but then realized that there is no need.

Reply:
I’ll add a comment. It was not obvious. 
I even had ‘Interlocked’, etc, before realizing it is single threaded and can be a lot simpler.